Waveform visualization

logiclrd opened the pull request:
This PR adds waveform visualization to the Info page.

In order to do this, the mixer code had to be updated to gather recent sample data. This needed to be as performant as possible, as I'm sure this is the hottest hot path in the codebase :slightly_smiling_face: I did my best in that regard. The buffers are statically-allocated, and the calculation and assignment of the values fits into the same spot as the VU calculation.

Also, in recognition of the possibility that some of the target architectures might yet have difficulty with the added CPU load (maybe? I don't know :slightly_smiling_face:), the feature can be turned off with `./configure --disable-waveformvis`. It is enabled by default as presently written.

I also tried making it join the waveform lines, but I actually like it more unjoined. I left that experiment in but `#ifdef`ed out, you can find it inside `page_info.c` under `#ifdef WAVEFORMVIS_JOINED`.

<img width="770" height="566" alt="image" src="https://github.com/user-attachments/assets/5396ee21-a919-4395-a343-217ca08d183a" />

logiclrd committed:
Added waveform visualization to the Info page:
- Added definition for --enable-waveformvis (default true) to configure.ac, and updated Makefile.am to transform it into -DENABLE_WAVEFORMVIS.
- Updated sndfile.h to conditionally define a static buffer for storing recent sample data for every voice as well as the output channels. Implemented the symbols in csndfile.c.
- Updated mixer.c to conditionally augment the generated mixing functions with code to store recent samples for each voice and for the mixing output. Did my best to make it reasonably toight.
- Updated struct info_window_type in page_info.c to include a recalculate callback, and to take a pointer to the info_window as a "this" pointer in all the callbacks. Updated the function definitions correspondingly. Updated all sites that initialize type to set type_def as well, and all call sites of callbacks to supply the this pointer.
- Updated struct info_window in page_info.c to track the calculated first_row, to store a pointer to the info_window_type definition associated with type, and to contain a vgamem_overlay for optional use by the implementation. Updated recalculate_windows to store the first_row value and call the recalculate callback if supplied.
- Added definition of new info_window_types, conditional on ENABLE_WAVEFORMVIS, that render the waveforms for selected channels/output.
- Updated the click callback to accept the mouse fine position as well as the character position. Updated info_handle_click correspondingly.
- Updated cfg_load_info_old and cfg_load_info in page_info.c to populate type_def alongside type.
- Updated cfg_load_info to have a fallback on a default layout instead of crashing the entire process if the configuration contains an unrecognized type name. :-) Factored default initialization to new function reset_window_layout.

logiclrd committed:
Oops :-) Apparently, "sndfile.h" was being satisfied with <sndfile.h> pulling from /usr/include/sndfile.h, which exists on my system but not on the build servers.

mrpapersonic commented on include/player/sndfile.h:
```diff
@@ -678,6 +681,25 @@ typedef struct song {
+#define RECENT_SAMPLE_BUFFER_SIZE NATIVE_SCREEN_WIDTH
+
+/* one for each voice, plus one for each output channel */
+extern int8_t recent_sample_buffer[(MAX_VOICES + 2) * RECENT_SAMPLE_BUFFER_SIZE];
```
The recent output buffer for each voice should be in the `song_voice_t` structure, and the two stereo output channels don't need to be there because they're already rendered out in the audio callback (which is how the oscilloscope visualizer works)

I think a circular buffer similar to events.c would work perfectly here

BTW, don't use globals in the replayer, since it's meant to be reentrable (see disko.c)

BTW BTW, we could use a flag to specify whether the mixer should render out this data, but I don't know if the possible performance increase is worth it

BTW BTW BTW this also has to be implemented for the AdLib stuff

logiclrd replied:
Fun fun fun :-D

So... is it okay to balloon `song_voice_t` by more than half a kilobyte?? I notice at least some parts of it seem to be really carefully laid out for cache line optimization and such. :-P

What if the current approach were used, but the buffer were in the owner of the `song_voice_t` (the `song_t`) rather than being static?

I also think a flag to specify whether the mixer should render out the data would likely be a performance _decrease_, because now it has to check that flag all the time. That's why I went with the approach of, "If you can't handle this, then we'll just rip it out entirely and it'll be like it was never there." :-)

logiclrd replied:
The existing output for the oscilloscope visualizer isn't guaranteed to be 8-bit signed. So, either it will need to be bent to this format, or the rendering code will need to be duplicated...

logiclrd replied:
Oh, I just remembered another reason I'm not storing it in `song_voice_t`. Every time you stop playback, it does a `memset(voices, 0, sizeof(voices))`. :-P That's in `_csf_reset`.

Do you think it makes more sense to still store it in `song_voice_t` and just make sure it doesn't get obliterated, or to store in a parallel structure?

mrpapersonic replied:
> So... is it okay to balloon song_voice_t by more than half a kilobyte?? I notice at least some parts of it seem to be really carefully laid out for cache line optimization and such. :-P

you can just put it directly out of the existing data, it's fine. That note was probably written over 20 years ago. If you're very hard pressed on keeping it optimized, put it directly before the `// Information not used in the mixer` comment.

mrpapersonic replied:
> The existing output for the oscilloscope visualizer isn't guaranteed to be 8-bit signed. So, either it will need to be bent to this format, or the rendering code will need to be duplicated...

The already-existing visualizers just bend the data to the format they need, and even then you can convert 16-bit signed to 8-bit signed with just a simple `rshift_signed(x, 8)`

mrpapersonic replied:
> Oh, I just remembered another reason I'm not storing it in `song_voice_t`. Every time you stop playback, it does a `memset(voices, 0, sizeof(voices))`. :-P That's in `_csf_reset`.
> 
> Do you think it makes more sense to still store it in `song_voice_t` and just make sure it doesn't get obliterated, or to store in a parallel structure?

shouldn't the visualization data get wiped when the song is stopped anyway?

logiclrd replied:
> I think a circular buffer similar to events.c would work perfectly here

It already is a circular buffer, just a small one. :-) Each voice gets `NATIVE_SCREEN_WIDTH` recent samples (which I chose because that is the widest you'd need with a one-to-one-mapping even if the visualization spanned the entire screen), and each voice stores an `oldest_recent_sample`. When new data is poured in, `oldest_recent_sample` is advanced, and when rendering, the buffer interpretation starts at `oldest_recent_sample` and wraps as needed.

mrpapersonic commented on player/mixer.c:
```diff
@@ -243,23 +255,35 @@ static inline SCHISM_ALWAYS_INLINE uint32_t avg_u32(uint32_t a, uint32_t b)
+# define CAPTURE_RECENT_SAMPLE
+#endif
+
+#define SNDMIX_STOREVUMETER_AND_RECENT_SAMPLE \
```
These should be separate

logiclrd replied:
Fair :-) In its earliest iteration, it shared the data. Then I came to realize that, actually, it's not the same data. Those graphs looked really weird :-)

mrpapersonic replied:
should be renamed back too

mrpapersonic commented on schism/page_info.c:
```diff
@@ -729,10 +735,133 @@ static void info_draw_note_dots(int base, int height, int active, int first_chan
+			vgamem_ovl_drawpixel(&window->overlay, x, y2, WAVEFORM_BOX_OUTLINE_COLOUR);
+
+		if ((chan >= MAX_VOICES) || current_song->voices[chan].current_sample_data) {
+			recent_samples = RECENT_SAMPLE_BUFFER(chan);
```
what's going on here? why not use the vgamem sample drawing functions?

logiclrd replied:
Good point :-)

logiclrd replied:
Hmm, the existing sample drawing functions can't draw to a sub-rectangle within an overlay, and overlays are allocated at character boundaries, not pixel boundaries. So, to use the existing functions, it would either be necessary to support having a separate overlay per channel, whose boundaries usually overlapped on a character, or to rewrite the sample rendering to allow a pixel (x1, y1)-(x2, y2) to be supplied by the caller.

What about an intermediate point, for starters, rehoming the per-pixel code into vgamem.c, so it may not be sharing the function but at least its in a more logical place??

logiclrd replied:
Also, both my code and the existing code assume you want a particular colour for your sample data, and they're not the same. :-)

mrpapersonic replied:
If anything, the code should just be changed to allow for any arbitrary coordinates (which means adding x,y,width,height params)

You can also just take in the color as another param

logiclrd committed:
Moved the recent sample buffer storage into song_t. Updated RECENT_SAMPLE_BUFFER to take csf to base its calculations on. Moved the use of RECENT_SAMPLE_BUFFER out of the hot path in mixer.c. Updated csf_allocate to allocate the memory for it and csf_free to free it.
Added labels to the channels in page_info:
- Added function vgamem_ovl_drawtext_halfwidth to render halfwidth characters into an overlay at any pixel (x, y).
- Updated struct info_window in page_info.c to track whether waveform labels should be hidden (so that the default with a zeroed struct is that they're visible).
- Updated info_page_handle_key to toggle the current window's hide_waveform_label on Alt-N.
- Updated info_draw_waveform in page_info.c to overlay labels when they're not hidden.
- Updated helptext/info-page to document Alt-N.
Moved the usage of CAPTURE_RECENT_SAMPLE out of SNDMIX_STOREVUMETER and into each of the use sites of SNDMIX_STOREVUMETER in mixer.c.
Updated page_info.c to capture the fact that the waveform visualization is visualizing voices, not channels:
- Added a second state variable selected_voice. Updated info_draw_waveform to use it instead of the caller-supplied first_channel (which comes from selected_channel).
- Updated struct info_window_type to capture whether the window type is showing voices or not. Updated the window_types initialization to set it to true for the window types showing voice data.
- Updated navigation in info_page_handle_key to update selected_voice first instead of selected_channel, except when pressing Left in a non-voice view, and to keep them synchronized.
- Updated info_draw_waveform to detect if the view would go past the end of the voices array and to clip the selection back so that the last voice displayed is at most MAX_VOICES.

logiclrd committed:
Fixed mixer.c and page_info.c to use the actual macro RECENT_SAMPLE_BUFFER_SIZE instead of the value it happens to have right now NATIVE_SCREEN_WIDTH.

logiclrd committed:
Fixed mixer.c to track the real voice behind a sample playing through voice_mix.
Updated page_info.c to additionally check if a voice has its volume set to 0 before deciding to render it.

logiclrd committed:
Updated snd_fm.c to store per-channel recent sample data like mixer.c. This also indirectly fixes the bug where samples coexisting with FM synthesis results in the audio output being connected to the 7th layer of hell.
Updated waveform rendering in page_info to check for voices rendering OPL synthesis.
Fixed a hypothetical buffer size issue with the channel number rendering in page_info.c.

logiclrd committed:
Added field opl_buffer_data to struct song in sndfile.h.
Updated _csf_reset in csndfile.c to zero the opl_buffer_data field and csf_destroy to free any allocation that has been assigned to it.
Updated snd_fm.c to allocate buffer space for opl_buffer_data on-demand instead of using variable-length stack allocation.

logiclrd committed:
Created new functions draw_sample_data_ex_[8,16,32] in vgamem.c/.h based on the sample drawing in page_info.cs and updated page_info to call draw_sample_data_ex_8 instead of doing it itself.
Updated draw_sample_data in vgamem.c to use draw_sample_data_ex_[8,16].
Updated info_draw_waveform in page_info to only draw main output when at least one channel is being mixed.

logiclrd committed:
Translated draw_sample_data_rect_[8,16,32] to use draw_sample_data_ex_[8,16,32]. Removed the DRAW_SAMPLE_DATA_VARIANT macro and its unused invocations.

logiclrd committed:
Replaced the separate implementations of draw_sample_data_rect_[8,16,32] with macro DRAW_SAMPLE_DATA_RECT_VARIANT in vgamem.c.

logiclrd committed:
Replaced the separate implementations of draw_sample_data_ex_[8,16,32] with macro DRAW_SAMPLE_DATA_EX_VARIANT in vgamem.c.

logiclrd committed:
Made definition of yy more traditional in DRAW_SAMPLE_DATA_EX_VARIANT in vgamem.c. It built for me, but it doesn't build on the OS X build servers. Shrug :-)

logiclrd committed:
Fixed up FM synthesis channel numbers:
- Updated Makefile.am to inject a macro FM_CHANNELS depending on whether OPL2 or OPL3 is enabled. OPL3 legitimately has 18 full independent stereo channels.
- Updated sndfile.h to use FM_CHANNELS as the size of the various per-channel OPL arrays.
- Updated OPLUpdateOne to use FM_CHANNELS for its meta buffer array size.
- Updated Fmdrv_Init, SetVoice and OPL_Reset in snd_fm.c to use FM_CHANNELS for vu_max and for all per-channel loops, including those with comments indicating confusion about 9 vs. 18 channels :-)
- Updated PortBases in snd_fm.c to specify 18 bases. I hope I did this right, looks like they were in groups of 3, advancing by 8 each group.

# Conflicts:
#	include/player/fmopl.h
#	include/player/sndfile.h
#	player/snd_fm.c

mrpapersonic commented on schism/vgamem.c:
```diff
@@ -768,70 +799,121 @@ void draw_vu_meter(int x, int y, int width, int val, int color, int peak)
-		int##bits##_t *data, uint32_t length, unsigned int inputchans, unsigned int outputchans) \
+#define DRAW_SAMPLE_DATA_EX_VARIANT(BITS, SCALE, SIGNARG_DEF, SIGN_FLIP_DEF, SIGN_FLIP_DO) \
+	void draw_sample_data_ex_##BITS(struct vgamem_overlay *r, int x1, int y1, int x2, int y2, \
+		int##BITS##_t *data, int offset, int stride, int advance, int length, SIGNARG_DEF \
```
can't `advance` be calculated with all the info already provided?

p.s. `offset`, `stride`, and `length` should all be `int32_t` or `uint32_t`

logiclrd replied:
`advance` is calculated just before the call to `draw_sample_data_ex`:
```
		int advance = _muldiv(sample->length * chans, 256, r->width * stride);
```
I don't think that information is available in the current parameter set. But, it's also a forced-inline function, and in the current implementation, it's only calculated in one place.

logiclrd replied:
Oh, also, `draw_data_sample_rect_BITS` also uses `draw_sample_data_ex_BITS` and uses a different `advance` calculation.

mrpapersonic commented on schism/page_info.c:
```diff
@@ -729,10 +740,147 @@ static void info_draw_note_dots(int base, int height, int active, int first_chan
+		if ((chan >= MAX_VOICES)
+			? current_song->mix_stat /* main mixer output -- is anything playing? */
+			: ((voice->current_sample_data && (voice->left_volume || voice->right_volume)) /* PCM */
+		 || (current_song->opl_from_chan[chan] >= 0))) /* FM */ {
```
`voice->flags & CHN_ADLIB` ?

logiclrd replied:
Mm, I'm not entirely convinced there isn't a 1-to-1 relationship between `CHN_ADLIB` being set and `opl_from_chan[chan]` being mapped, but maybe it doesn't get cleared or something. Okeydoke :-)

mrpapersonic commented on player/csndfile.c:
```diff
@@ -34,6 +34,20 @@
 #include "mem.h"
 
+#ifdef ENABLE_WAVEFORMVIS
+int8_t recent_sample_buffer[(MAX_VOICES + 2) * RECENT_SAMPLE_BUFFER_SIZE];
```
unused externs

logiclrd replied:
:+1:

mrpapersonic commented on player/snd_fm.c:
```diff
@@ -257,43 +254,102 @@ void Fmdrv_Mix(song_t *csf, uint32_t count)
+	else {
+		if (!csf->opl_buffer_data) {
+			SCHISM_RUNTIME_ASSERT(sz <= MIXBUFFERSIZE * 2, "Fmdrv_Mix can only process up to MIXBUFFERSIZE samples at a time, caller requested more");
+			csf->opl_buffer_data = mem_calloc(OPL_CHANNELS, MIXBUFFERSIZE * 2 * sizeof(int32_t));
```
since this is constant size, it can be moved into the csf init function

logiclrd committed:
changes for PR feedback
* eliminated static symbols for recent sample data for output samples. moved oldest_recent_output_sample into struct song and eliminated accessor functions.
* updated info_draw_waveform in page_info.c to check explicitly that the voice is flagged CHN_ADLIB in addition to checking for an OPL channel mapping
* made parameters in draw_sample_data_ex_BITS and draw_sample_data_rect_BITS in vgamem.c/.h that assume they have 32 bits of storage into explicit 32-bit values
* moved allocation of opl_buffer_data into csf_allocate, alongside the allocation of recent_sample_buffer
also, updated code that resets voices to also reset recent_sample_buffer
